src/config.rs

```diff
@@ -510,26 +588,90 @@ impl Cfg {
         Ok(None)
     }
 
+    fn parse_override_file<S: AsRef<str>>(contents: S) -> Result<Option<OverrideFile>> {
+        let contents = contents.as_ref();
+
+        if contents.lines().any(|line| line.starts_with("[toolchain]")) {
```

Comment:
@rbtcollins @kinnison following our discussion on Discord, I've applied this suggestion by avoiding to manually parse the TOML format, and showing the TOML parse error in addition to the "unknown channel" error if parsing as legacy format yields an invalid toolchain name. I've not used serde-plain because (correct me if I'm wrong) it would not respect the old behavior given that lines following the first one were simply ignored. I believe that same reason prevents trying to parse as legacy format before TOML.

I find this solution is not ideal, but it works. There's still one problem though, we can't control the error message for old rustups, as the contents of the first line could be anything. We can't require the first line to be anything specific if we want to avoid parsing TOML manually (which I think is wise advice, as e.g. `[ toolchain ]` is valid TOML).

All in all, I strongly feel that having a specific `rustup.toml` would get rid of all these problems, if I may propose it again at this stage (one last time, I promise :)). We can prevent having both files by erroring out so diverging contents should not be a concern. The only drawback I can think of is that it could make the integration tests slower, which I believe is currently a concern.

Let me know your thoughts.

Reply:
the failure mode though is that older rustups will then behave as though there is no rustup file at all, which will be mysterious.

Reply:
I'd rather risk unusual but failing behaviour from an old `rustup` over silent miscompilation with the wrong toolchain, or unusual errors as a result.  Thus no, `rustup.toml` is not going to be a thing at this point.

The way I see it, a reasonable approach would be:

1. An empty `rust-toolchain` file is a specific error, perhaps "Empty `rust-toolchain` file detected.  Please remove it, or else specify the desired toolchain properties in the file".
2. A single-line `rust-toolchain` file is to be treated as the old-form, even if it's `[toolchain]` or similar, the resultant error will probably be obvious
3. Any other file content should be parsed-as-toml and errors reported as appropriate.

Reply:
Alright, I totally understand the concerns about the old rustups, I'll implement the change as described.

Just to explain the way I understood it, is that pre-`rust-toolchain` rustups were ignoring the file too.

Reply:
Fair enough, but that was 3 years ago (June 2017), and we can't do anything about those versions: if someone on RHEL or Ubuntu 1604 or whatnot running an LTS version of rustup tries to build, they will get unexpected results, and we may have to support them; but we are at least not creating new silent failure modes: any newer rustup will fail verbosely.